### ISSUE TITLE
Tracking issue: Directives Hydration ⚛️

### ISSUE THREAD

**Issue by luisherranz** [luisherranz/block-hydration-experiments#64]
**We are now actively working on the [WordPress Directives Plugin Tracking issue](https://github.com/WordPress/block-hydration-experiments/issues/80) instead** ➡️

---

This issue aims to list and keep track of the tasks related to the **Directives Hydration experiment**.

- https://github.com/WordPress/block-hydration-experiments/issues/60

_For the Custom Elements Hydration experiment, please check the [Tracking Issue: Custom Elements Hydration 🧩](https://github.com/WordPress/block-hydration-experiments/issues/39)._

The main branch of this experiment is `main-directives-hydration`. Make sure you select it when opening a pull request if it is related to this experiment.

Please, proactively comment on this issue whenever you start working on a new task, get blocked, or finish a task, sharing as much detail and information as possible. Thanks!!

# Experiments

- https://github.com/WordPress/gutenberg/pull/44034
- https://github.com/woocommerce/woocommerce-blocks/pull/7187
- https://github.com/WordPress/gutenberg/pull/44509
- https://github.com/DAreRodz/js-framework-benchmark/pull/1

# Done

- https://github.com/WordPress/block-hydration-experiments/pull/65
- https://github.com/WordPress/block-hydration-experiments/pull/66
- https://github.com/WordPress/block-hydration-experiments/pull/68
- https://github.com/WordPress/block-hydration-experiments/pull/70
- https://github.com/WordPress/block-hydration-experiments/pull/71
- https://github.com/WordPress/block-hydration-experiments/pull/72
- https://github.com/WordPress/block-hydration-experiments/pull/77
- https://github.com/WordPress/block-hydration-experiments/pull/74
- https://github.com/WordPress/block-hydration-experiments/pull/78

# Next

# On hold
- https://github.com/WordPress/block-hydration-experiments/issues/75

**Comment by luisherranz** [comment]
I've opened the first PR to prepare the project to add the initial version of the hydration (co-authored with @darerodz):

- https://github.com/WordPress/block-hydration-experiments/pull/65

**Comment by DAreRodz** [comment]
#65 is approved and merged. 🎉  I'll update the tracking issue's description to reflect that.

Aso, I've started with a PR to hydrate block components inside a single vDOM tree:

- https://github.com/WordPress/block-hydration-experiments/pull/66

**Comment by luisherranz** [comment]
I've made a proof of concept of the vdom-based client-side navigation on the Query Loop block:

- https://github.com/WordPress/gutenberg/pull/44034

**Comment by luisherranz** [comment]
I've made some more progress on the experiment, replacing the imperative event listeners with a `wp-client-navigation` directive and adding a small client router that caches the HTML and can prefetch pages: 

- https://github.com/WordPress/gutenberg/pull/44034#issuecomment-1244105205

So far so good!

**Comment by DAreRodz** [comment]
The "block wiring" is almost complete; I marked #66 ready for review. Still, I have intentionally left the block context to implement it in a different PR and test the [preact's option hook API](https://preactjs.com/guide/v10/options/).

- #66 

There's also a problem during hydration, because some elements are being regenerated. We need to investigate that. 🕵️ 

EDIT: the hydration bug was found. I left a [note](https://github.com/WordPress/block-hydration-experiments/pull/66#discussion_r969799786) in the bugfix.

**Comment by DAreRodz** [comment]
I refactored `toVdom()`, removed `visitor.js` and added directives support. In addition, most of the code that was inside `visitor.js` has been moved to directives.

- https://github.com/WordPress/block-hydration-experiments/pull/68

**Comment by luisherranz** [comment]
David is working on `Suspense` and `ErrorBoundaries` in this PR:

- https://github.com/WordPress/block-hydration-experiments/pull/70

**Comment by luisherranz** [comment]
Destroying `<scripts>` was causing problems with the initial hydration, so @darerodz and I fixed that here: 

- https://github.com/WordPress/block-hydration-experiments/pull/71

**Comment by luisherranz** [comment]
I've started another client-side navigation experiment for WooCommerce's Product Query block. This one is replacing only the inner children of the Product Query block instead of the whole document.

- https://github.com/woocommerce/woocommerce-blocks/pull/7187

**Comment by DAreRodz** [comment]
I've removed the `<wp-block>` wrapper and everything seems to work fine. 🙌 

The way to append the wrapper attributes is a bit hacky, TBH. We can give it a thought and change that with something else.

- https://github.com/WordPress/block-hydration-experiments/pull/72

**Comment by luisherranz** [comment]
I'll be working on this today:

- Issue: https://github.com/WordPress/block-hydration-experiments/issues/73
- PR: https://github.com/WordPress/block-hydration-experiments/pull/74

**Comment by luisherranz** [comment]
@darerodz, as part of https://github.com/WordPress/block-hydration-experiments/issues/73 do you mind if I refactor the directives code a bit to make it work with `wp-directive` instead of `data-wp-block-directive`? Those `data-wp-block-` attributes were internals, but now that we're going to expose them to the devs, I'd rather go with `wp-`.

**Comment by DAreRodz** [comment]
> do you mind if I refactor the directives code a bit to make it work with `wp-directive` instead of `data-wp-block-directive`?

Of course I don't mind! Actually, I thought of doing the same, hehe. 😄

**Comment by DAreRodz** [comment]
I've written an update on the lazy hydration techniques issue:

- https://github.com/WordPress/block-hydration-experiments/issues/75#issuecomment-1256505760

TL;DR: all the approaches we tried so far were unsuccessful, although we have some ideas we're going to explore next week.

**Comment by DAreRodz** [comment]
We decided to put #75 on hold as mentioned in https://github.com/WordPress/block-hydration-experiments/issues/75#issuecomment-1257669198, and continue with other tasks.

Also, I've created an issue for removing the `wp-inner-blocks` wrapper, which is what I'm going to work on.

- https://github.com/WordPress/block-hydration-experiments/issues/76

**Comment by luisherranz** [comment]
@ockham and I recorded a video summary of the progress with this.

<a href="https://www.loom.com/share/fa2dc7c81cd14c6ebef6bc2d10863979">
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/fa2dc7c81cd14c6ebef6bc2d10863979-with-play.gif">
  </a>

https://www.loom.com/share/fa2dc7c81cd14c6ebef6bc2d10863979

EDIT: @santosguillamot [discovered what is the `cst` query](https://github.com/WordPress/gutenberg/pull/44034#issuecomment-1259222564).

**Comment by luisherranz** [comment]
#74 should be ready for review, although I just noticed a problem with the block context that I want to investigate.

- https://github.com/WordPress/block-hydration-experiments/issues/74

**Comment by DAreRodz** [comment]
I've removed the `wp-inner-blocks` wrapper in https://github.com/WordPress/block-hydration-experiments/pull/77. Now, a `wp-inner-block` attribute is included in each inner block instead, which is used to compute all inner blocks. The PR is ready to review.

- https://github.com/WordPress/block-hydration-experiments/pull/77

**Comment by luisherranz** [comment]
@santosguillamot, @darerodz and I have started this new experiment to see if we can replicate the behavior of the Navigation block with directives hydration:

- https://github.com/WordPress/gutenberg/pull/44509

It's based on @santosguillamot's [initial exploration using Alpine directives](https://github.com/WordPress/gutenberg/pull/44289).

**Comment by luisherranz** [comment]
I just merged #77. I'll resolve the conflicts on #74 and fix its bug.

**Comment by luisherranz** [comment]
https://github.com/WordPress/block-hydration-experiments/pull/74 is ready for review:

- https://github.com/WordPress/block-hydration-experiments/pull/74

EDIT: Merged! 🎉

**Comment by luisherranz** [comment]
I think we've finished pretty much here. 

@darerodz, is there something else we should add?

**Comment by DAreRodz** [comment]
In a pair-programming session with @c4rl0sbr4v0, we implemented a way to pause hydration if View components are not registered yet.

- https://github.com/WordPress/block-hydration-experiments/pull/78

**Comment by luisherranz** [comment]
Merged 🙂

**Comment by DAreRodz** [comment]
> I think we've finished pretty much here.
> 
> DAreRodz, is there something else we should add?

Uhm, I'm not sure what else we could add. 😄

Maybe we can test if the directives hydration approach could get along with third-party blocks using other libraries for interactivity. As long as they don't interact with our VDOM, the hydration is supposed to work fine, although we haven't tested yet if that assumption is correct.

**Comment by luisherranz** [comment]
I've made a [video summary of the "Replicate Navigation block using vDOM-based directives" experiment](https://github.com/WordPress/gutenberg/pull/44509#issuecomment-1263667869).

**Comment by DAreRodz** [comment]
I've started working on a benchmark to measure WP directives performance.

- https://github.com/DAreRodz/js-framework-benchmark/pull/1

I placed a link in the description under **Experiments**. Feel free to move it if you feel there's a better place for it.

**Comment by luisherranz** [comment]
We've moved our work to the [WP Directives Plugin](https://github.com/WordPress/block-hydration-experiments/issues/80) branch and Tracking Issue because we are now trying to build a plugin that people can install to test this out.

**Comment by luisherranz** [comment]
We have renamed this to Directives Hydration to make more explicit the difference between this and [the Custom Elements Hydration](https://github.com/WordPress/block-hydration-experiments/issues/39).

**Comment by luisherranz** [comment]
Closed in favor of:

- https://github.com/WordPress/block-hydration-experiments/issues/80
